File: spiral-matrix.js
// first row ned to rpint as it array[0] 
// then last column that would be array[1-length][length-1]


let spiralOrder = function(matrix, result = []) {
    if (matrix.length === 0) return result;

    // Step 1: Top row (left to right)
    result.push(...matrix.shift());

    // Step 2: Right column (top to bottom)
    for (let i = 0; i < matrix.length; i++) {
        if (matrix[i].length > 0) {
            result.push(matrix[i].pop());
        }
    }

    // Step 3: Bottom row (right to left)
    if (matrix.length > 0) {
        result.push(...(matrix.pop() || []).reverse());
    }

    // Step 4: Left column (bottom to top)
    for (let i = matrix.length - 1; i >= 0; i--) {
        if (matrix[i].length > 0) {
            result.push(matrix[i].shift());
        }
    }

    // Recursive call
    return spiralOrder(matrix, result);
}

// Example usage:
console.log(spiralOrder([[1],[2],[3],[4],[5],[6],[7],[8],[9],[10]])); // Output: [1, 2, 3, 6, 9, 8, 7, 4, 5]

File: spiral-matrix.py

def spiralOrder(self, matrix):
    """
    :type matrix: List[List[int]]
    :rtype: List[int]
    """
    result=[]
    top,left=0,0
    bottom,right=len(matrix)-1,len(matrix[0])-1
    while top<=bottom and left<=right:
        # for the first row
        for i in range(left,right+1):
            result.append(matrix[top][i])
        top+=1
        # for the last column
        for i in range(top,bottom+1):
            result.append(matrix[i][right])
        right-=1
        if top<=bottom:
            # for the last row
            for i in range(right,left-1,-1):
                result.append(matrix[bottom][i])
            bottom-=1
        if left<=right:
            # for the first column
            for i in range(bottom,top-1,-1):
                result.append(matrix[i][left])
            left+=1
    return result

# Example usage:
if __name__ == "__main__":
    matrix = [
        [1, 2, 3,12],
        [4, 5, 6,11],
        [7, 8, 9,10]
    ]
    print(spiralOrder(None, matrix))  # Output: [1, 2, 3, 6, 9, 8, 7, 4, 5]